# rna-seq-blood-mice-toxoplasma-2025
## Author 
Vittoria Mungai 

## Description 
Bulk RNA-sequencing and differential expression analysis of blood samples from mice. The mice are divided in two groups: one is infected with toxoplasma and the other is uninfected control groups. 

## Sample data 
The fastq files from the blood samples used are stored in the directory: `/data/courses/rnaseq_course/toxoplasma_de/reads_Blood`. 



### Structure of the workflow 
Starting from the fastq files of the samples the workflow is divided in 6 main steps: 
1. Quality checks
2. Map reads to the reference genome
3. Count the number of reads per gene
4. Exploratory data analysis
5. Differential expression analysis
6. Overrepresentation analysis

#### 1. Quality checks 
Before doing the fastqc quality check an intermediate .tsv file was created with the script "generate_sampleslist.sh" into the directory "rna_seq_projects/rna-seq-blood-mice-toxoplasma-2025/results/fastqc/intermediate_results" to list all the names and fastq files of the samples. The name of the file is "sampleslist.tsv", it is fixed because it was used by the script "run_fastqc_with_slum_array.sh"

For the quality check the following scripts (in the `scripts` directory) were run:
1. in the folder `/fastqc`:
    - `generate_sampleslist.sh` to generate the file `sampleslist.tsv` with the list of all the samples and the corrispective fastq files. 
    - `run_fastqc_with_slum_array.sh` to generate the fastqc files for each fragment. 
2. in the folder `/multiqc`: 
    - `run_multiqc.sh` to generate the multiqc report and have a overview of all the fragments in one single report. 
3. in the folder `/fastp_trimming`: 
    - `trim_adapters_fastp.sh` to trim the fastq files. This script uses the file `results/fastqc_after_trimming/intermediate_results/sampleslist.tsv` generated by the script `/scripts/fastqc/generate_sampleslist.sh`
    in the folder `/fastqc`: 
    - `run_fastqc_with_slum_array.sh` to generate the fastqc files for each trimmed sample 
4. in the folder `hista2`: 
    - `run_hista2_index_files.sh`
    - `map_reads_reference_genome.sh`
    in the folder `shared`: 
    - `generate_sam_bam_lists.sh`to generate the file "sam_bam_files_list.tsv" needed for the following 2 scripts 
    - `convert_sam_to_bam.sh` it uses the file "sam_bam_files_list.tsv"
    - `sort_bam_files.sh` it uses the file "sam_bam_files_list.tsv"
    in the folder `shared`: 
    - `generate_sorted_bam_list.sh` to generate le list of the sorted bam files that will be used to use the slurm array in the following script 
    - `index_sorted_bam_files.sh` it uses the file "sorted_bam_list.tsv"





