# rna-seq-blood-mice-toxoplasma-2025
## Author 
Vittoria Mungai 

## Description 
Bulk RNA-sequencing and differential expression analysis of blood samples from mice. The mice are divided in two groups: one is infected with toxoplasma and the other is uninfected control groups. 

## Sample data 
The fastq files from the blood samples used are stored in the directory: `/data/courses/rnaseq_course/toxoplasma_de/reads_Blood/`. It's possible to access the folder directly when needed of create a logical link with the command `ln -s [name_link] [name_original_file]`



### Structure of the workflow 
Starting from the fastq files of the samples the workflow is divided in 6 main steps: 
1. 
2. Quality checks
3. Map reads to the reference genome
4. Count the number of reads per gene
5. Exploratory data analysis
6. Differential expression analysis
7. Overrepresentation analysis

:note: 
- In scripts where parameters are requested there is the command with all the arguments for each execution
- If a scripts access a file which needs to be generated, a explicit comment in the script will tell you. 
- The output of the script are all in the folder results. ALthough, further details are given below. 

#### 1. Accessing the raw data from the cluster 
The fastq files for the Blood sample are stored in the directory `/data/courses/rnaseq_course/toxoplasma_de

#### 2. Quality checks 

For the quality check the listed scripts (in the `/scripts` directory) were run in the following order:
1. in the folder `/shared`:
    - `generate_sampleslist.sh` to generate the file `rna_seq_projects/rna-seq-blood-mice-toxoplasma-2025/results/fastqc/intermediate_results/sampleslist.tsv` with the list of all the samples and the corrispective fastq files. 
    - `run_fastqc_with_slum_array.sh` to generate the fastqc files for each fragment. 
2. in the folder `/multiqc`: 
    - `run_multiqc.sh` to generate the multiqc report and have a overview of all the fragments in one single report. 
3. in the folder `/fastp_trimming`: 
    - `trim_adapters_fastp.sh` to trim the fastq files. This script uses the file `results/fastqc_after_trimming/intermediate_results/sampleslist.tsv` generated by the script `/scripts/fastqc/generate_sampleslist.sh`
    in the folder `/fastqc`: 
    - `run_fastqc_with_slum_array.sh` to generate the fastqc files for each trimmed sample 
 
#### 3. Map reads to the reference genome
1.  Produce all required index files for Hisat2
    in the folder `hisat2`: 
    - `run_hisat2_index_files.sh`
2. For each sample separately, map the reads to the reference genome using Hisat2. The correct strandedness setting for this library prep protocol is RF.
3. Convert the resulting sam files to bam format using Samtools 
4. Sort the bam files by genomic coordinates using Samtools
    - `map_reads_convert_sort_bam.sh` the steps 2, 3, and 4 are achieved with this script. The output will be only the sorted bam file. 
5. Index the coordinate sorted bam files using Samtools 
    in the folder `shared`: 
    - `generate_sorted_bam_list.sh` to generate le list of the sorted bam files that will be used to use the slurm array in the script: `index_sorted_bam_files.sh`
    - `index_sorted_bam_files.sh` it uses the file "sorted_bam_list.tsv" to index the bam files 

#### 4. Count the number of reads per gene
1. With `featureCounts` a table of counts is generated containing the number of reads per gene in each sample. 
    in folder `feature_counts`: 
    - `feature_counts.sh` will generate the table of counts and the counting summary. 
    - `multiqc_feature_counts.sh` will generate a .html report of featureCounts. 




